kafka/src/main/java/com/linecorp/armeria/server/logging/kafka/KafkaLogWriter.java

```diff
@@ -0,0 +1,166 @@
+     *                               a {@link RequestLog} when writing request log.
+     *                               The {@link Function} is allowed to return {@code null}
+     *                               to leave the record key unspecified.
+     * @param responseLogKeyExtractor a {@link Function} that extracts a {@code V}-typed record value from
+     *                                a {@link RequestLog} when writing response log.
+     *                                The {@link Function} is allowed to return {@code null}
+     *                                to skip logging for the given {@link RequestLog}.
+     * @param requestLogLevelMapper a {@link RequestLogLevelMapper} to use when mapping the log level
+     *                              of request logs.
+     * @param responseLogLevelMapper a {@link ResponseLogLevelMapper} to use when mapping the log level
+     *                               of response logs.
+     * @param logFormatter a {@link LogFormatter} which converts a {@link RequestOnlyLog} or {@link RequestLog}
+     *                     into a log message
+     */
+    KafkaLogWriter(Producer<K, String> producer, String topic, Logger logger,
+                   Function<? super RequestOnlyLog, ? extends @Nullable K> requestLogKeyExtractor,
+                   Function<? super RequestLog, ? extends @Nullable K> responseLogKeyExtractor,
+                   RequestLogLevelMapper requestLogLevelMapper, ResponseLogLevelMapper responseLogLevelMapper,
+                   LogFormatter logFormatter) {
+        this.producer = requireNonNull(producer, "producer");
+        this.topic = requireNonNull(topic, "topic");
+        this.logger = requireNonNull(logger, "logger");
+        this.requestLogKeyExtractor = requireNonNull(requestLogKeyExtractor, "requestLogKeyExtractor");
+        this.responseLogKeyExtractor = requireNonNull(responseLogKeyExtractor, "responseLogKeyExtractor");
+        this.requestLogLevelMapper = requireNonNull(requestLogLevelMapper, "requestLogLevelMapper");
+        this.responseLogLevelMapper = requireNonNull(responseLogLevelMapper, "responseLogLevelMapper");
+        this.logFormatter = requireNonNull(logFormatter, "logFormatter");
+        ShutdownHooks.addClosingTask(producer, "producer for KafkaLogWriter");
+    }
+
+    @Override
+    public void logRequest(RequestOnlyLog log) {
+        requireNonNull(log, "log");
+        LogLevel requestLogLevel = requestLogLevelMapper.apply(log);
+        if (requestLogLevel == null) {
+            if (!warnedNullRequestLogLevelMapper) {
+                warnedNullRequestLogLevelMapper = true;
+                logger.warn("requestLogLevelMapper.apply() returned null; using default log level");
+            }
+            requestLogLevel = DEFAULT_REQUEST_LOG_LEVEL;
+        }
+        if (requestLogLevel.isEnabled(logger)) {
+            try (SafeCloseable ignored = log.context().push()) {
```

Comment:
What do you think of adding two predicates to determine whether the log should be written? It would be weird to use `LogLevel` to filter the logs sent to Kafka. `LogLevel` is related to the logging system such as Log4j or Logback.
Users may want to log failed requests by default. We can refer to `SuccessFunction.ofDefault()` to implement the default predicate.
https://github.com/line/armeria/blob/c8d1f5ef142f65fd79bd01b87d8155d220c9d0b9/core/src/main/java/com/linecorp/armeria/common/SuccessFunction.java#L86-L98

Additionally, it might be better to log the request and response together so that users don't have to correlate them. `LogFormatter.format()` seems necessary as a new API to format a request and response as a single output.
```json
{
  "request": {
    ...
  },
  "response": {
    ...
  }
}
```

```java
boolean shouldLogRequest = requestLogPreciate.test(log);
boolean shouldLogResponse = responseLogPreciate.test(log);

String formatted;
if (responseHasCause || (shouldLogRequest && shouldLogResponse)) {
  formatted = logFormatter.format(log);
} else if (shouldLogRequest) {
  formatted = logFormatter.formatRequest(log);
} else {
  formatted = logFormatter.formatResponse(log);
}
```

Reply:
> What do you think of adding two predicates to determine whether the log should be written? It would be weird to use LogLevel to filter the logs sent to Kafka.

Seems good.

> Additionally, it might be better to log the request and response together so that users don't have to correlate them.

Then, does it mean that `KafkaLogWriter#logRequest` does nothing?

Reply:
> Then, does it mean that KafkaLogWriter#logRequest does nothing?

Because many logs in serval servers could be collected in a Kafka topic, a unified log will be easier to distinguish. However, I'm not 100% sure if it is a better suggestion. Let's listen to other folks' options more.

Reply:
It makes sense to me.

Reply:
That sounds good to me as well.
We should also have the corresponding API to the `LogWriter`.
```java
public interface LogWriter {
  void write(RequestLog log);
  void logRequest(RequestOnlyLog log); // Should we rename like `writeRequestLog`.
  void logResponse(RequestLog log);
}
```

Reply:
Do we need to declare `write` method?
I think `logRequest` and `logResponse` would be sufficient.

Reply:
If you need to serialize both request and response into one document, then we need `write(RequestLog)`.

Reply:
Yes as @trustin said. 😉

Reply:
> // Should we rename like `writeRequestLog`?

Personally, I prefer `void log(RequestLog log)` for consistency.